internal/parse/parser.go

```diff
@@ -111,14 +119,33 @@ func (p *Parser) add(part queryPart) {
 
 // Parse takes an input string and parses the input and output parts. It returns
 // a pointer to a ParsedExpr.
-func (p *Parser) Parse(input string) (*ParsedExpr, error) {
+func (p *Parser) Parse(input string) (expr *ParsedExpr, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot parse expression: %s", err)
+		}
+	}()
 	p.init(input)
+
 	for {
 		p.partStart = p.pos
-		if p.pos == len(p.input) {
+
+		if ip, ok, err := p.parseInputExpression(); err != nil {
+			return nil, err
+		} else if ok {
+			p.add(ip)
+
+		} else if sp, ok, err := p.parseStringLiteral(); err != nil {
```

Comment:
The formatting here feels slightly hard to read and follow. Have you tried breaking this down into independent blocks (no else) with their own continue/break/return?  Half of them are already returning or breaking, which means they'll interrupt the flow regardless of the else.